### PR TITLE
fix(agent): recover direct skill tool calls

### LINKED ISSUE #4825: No ToolCallback found for tool name
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Current Behavior

ReactAgent 中注册了技能 技能名称**web-composite-search**，
[SKILL.md](https://github.com/user-attachments/files/30162727/SKILL.md) 然后tools添加了调用了PYTHON+MCPTOOLS。 
**版本spring-ai  1.1.2 , spring-ai-alibaba 1.1.2.3**

本来直接调用搜索技能执行里面的脚本命令。但是现在查询到技能把技能名作为TOOL名，调用TOOL的方法，造成 No ToolCallback found for tool name: web-composite-search。但是我把MCP的TOOL去掉。就没这个问题，我核查了MCO里面没有工具名是web-composite-search 。 唯一可能性，是MD文件里面描述有工具两个字 比如：  智能多引擎聚合搜索工具

     SkillRegistry skillRegistry = FileSystemSkillRegistry.builder().projectSkillsDirectory(skillpath).build();
     this.skillsAgentHook = SkillsAgentHook.builder().skillRegistry(skillRegistry).build();

private ToolCallback[] buildAgenticTools() {
        List<ToolCallback> tools = new ArrayList<>();
        tools.add(PythonTool.createPythonToolCallback(PythonTool.DESCRIPTION));
        // **如果这里去掉就可以正常调用。否则技能会调用到 ToolCallResponse executeToolCallWithInterceptors  出现 No ToolCallback found for tool name: web-composite-search** 
        if (mcpToolProvider != null) {
            tools.addAll(mcpToolProvider.getToolCallbacks());
        }
        return tools.toArray(new ToolCallback[0]);
    }

  ToolCallback[] agenticTools = buildAgenticTools(); // python + 城市/用户记忆 + 数据库驱动的原生 MCP 工具（动态）
            ReactAgent agent = ReactAgent.builder()
                    .name("agentic_" + key)
                    .model(model)
                    .tools(agenticTools)
                    .hooks(List.of(skillsAgentHook, shellHook, new MessageTrimmingHook(),
                            new MessageDataBaseHook(),
                            new RAGAgentHook(milvusHybridSchemaService),
                            SummarizationHook.builder()
                                    .model(model)
                                    .maxTokensBeforeSummary(8000)
                                    .messagesToKeep(20)
                                    .build()))
                    .interceptors(new RAGContextInterceptor())
                    .saver(mysqlSaver)
                    .build();
 

java.lang.IllegalStateException: No ToolCallback found for tool name: web-composite-search
	at com.alibaba.cloud.ai.graph.agent.node.AgentToolNode.lambda$executeToolCallWithInterceptors$0(AgentToolNode.java:547) ~[spring-ai-alibaba-agent-framework-1.1.2.3.jar:1.1.2.3]
	Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException: 
Assembly trace from producer [reactor.core.publisher.MonoCompletionStage] :
	reactor.core.publisher.Mono.fromFuture(Mono.java:629)
	com.alibaba.cloud.ai.graph.executor.NodeExecutor.executeNode(NodeExecutor.java:128)
Error has been observed at the following site(s):
	*____________________________________________Mono.fromFuture ⇢ at com.alibaba.cloud.ai.graph.executor.NodeExecutor.executeNode(NodeExecutor.java:128)
	|_                                          Mono.flatMapMany ⇢ at com.alibaba.cloud.ai.graph.executor.NodeExecutor.executeNode(NodeExecutor.java:129)
	*____________________________________________Mono.fromFuture ⇢ at com.alibaba.cloud.ai.graph.CompiledGraph.lambda$streamFromInitialNode$0(CompiledGraph.java:594)
	|_                                           Mono.onErrorMap ⇢ at com.alibaba.cloud.ai.graph.CompiledGraph.lambda$streamFromInitialNode$0(CompiledGraph.java:594)
	|_                                                 Mono.flux ⇢ at com.alibaba.cloud.ai.graph.CompiledGraph.lambda$streamFromInitialNode$0(CompiledGraph.java:594)
	*__MonoCompletionStage$MonoCompletionStageSubscription.apply ⇢ at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:934)
	*_______________________________________________Flux.flatMap ⇢ at com.alibaba.cloud.ai.graph.CompiledGraph.streamFromInitialNode(CompiledGraph.java:585)
	|_                                               Flux.filter ⇢ at com.kk.agent.controller.ChatAgentController.agentstream(ChatAgentController.java:224)
	|_                                                  Flux.map ⇢ at com.kk.agent.controller.ChatAgentController.agentstream(ChatAgentController.java:225)
	|_                                               Flux.filter ⇢ at com.kk.agent.controller.ChatAgentController.agentstream(ChatAgentController.java:226)
	|_                                             Flux.doOnNext ⇢ at com.kk.agent.controller.ChatAgentController.agentstream(ChatAgentController.java:228)
	|_                                               Flux.filter ⇢ at com.kk.agent.controller.ChatAgentController.agentstream(ChatAgentController.java:240)
	|_                                                  Flux.map ⇢ at com.kk.agent.controller.ChatAgentController.agentstream(ChatAgentController.java:241)
	|_                                         Flux.contextWrite ⇢ at org.springframework.web.servlet.mvc.method.annotation.ReactiveTypeHandler$ContextSnapshotHelper.writeReactorContext(ReactiveTypeHandler.java:596)
Original Stack Trace:
		at com.alibaba.cloud.ai.graph.agent.node.AgentToolNode.lambda$executeToolCallWithInterceptors$0(AgentToolNode.java:547) ~[spring-ai-alibaba-agent-framework-1.1.2.3.jar:1.1.2.3]
		at com.alibaba.cloud.ai.graph.agent.node.AgentToolNode.executeToolCallWithInterceptors(AgentToolNode.java:577) ~[spring-ai-alibaba-agent-framework-1.1.2.3.jar:1.1.2.3]
		at com.alibaba.cloud.ai.graph.agent.node.AgentToolNode.executeToolCallWithInterceptors(AgentToolNode.java:511) ~[spring-ai-alibaba-agent-framework-1.1.2.3.jar:1.1.2.3]
		at com.alibaba.cloud.ai.graph.agent.node.AgentToolNode.executeToolCallsSequential(AgentToolNode.java:230) ~[spring-ai-alibaba-agent-framework-1.1.2.3.jar:1.1.2.3]
		at com.alibaba.cloud.ai.graph.agent.node.AgentToolNode.apply(AgentToolNode.java:171) ~[spring-ai-alibaba-agent-framework-1.1.2.3.jar:1.1.2.3]
		at com.alibaba.cloud.ai.graph.action.AsyncNodeActionWithConfig.lambda$node_async$0(AsyncNodeActionWithConfig.java:41) ~[spring-ai-alibaba-graph-core-1.1.2.3.jar:1.1.2.3]
		at com.alibaba.cloud.ai.graph.executor.NodeExecutor.executeNode(NodeExecutor.java:125) ~[spring-ai-alibaba-graph-core-1.1.2.3.jar:1.1.2.3]
		at com.alibaba.cloud.ai.graph.executor.NodeExecutor.execute(NodeExecutor.java:87) ~[spring-ai-alibaba-graph-core-1.1.2.3.jar:1.1.2.3]
		at com.alibaba.cloud.ai.graph.executor.MainGraphExecutor.execute(MainGraphExecutor.java:108) ~[spring-ai-alibaba-graph-core-1.1.2.3.jar:1.1.2.3]
		at com.alibaba.cloud.ai.graph.executor.NodeExecutor.lambda$handleActionResult$0(NodeExecutor.java:209) ~[spring-ai-alibaba-graph-core-1.1.2.3.jar:1.1.2.3]
		at reactor.core.publisher.FluxDefer.subscribe(FluxDefer.java:46) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.Flux.subscribe(Flux.java:8891) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onComplete(FluxConcatArray.java:238) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxConcatArray.subscribe(FluxConcatArray.java:79) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.Flux.subscribe(Flux.java:8891) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.MonoFlatMapMany$FlatMapManyMain.onNext(MonoFlatMapMany.java:196) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.MonoCompletionStage$MonoCompletionStageSubscription.apply(MonoCompletionStage.java:121) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.MonoCompletionStage$MonoCompletionStageSubscription.apply(MonoCompletionStage.java:67) ~[reactor-core-3.7.8.jar:3.7.8]
		at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:934) ~[na:na]
		at java.base/java.util.concurrent.CompletableFuture.uniHandleStage(CompletableFuture.java:950) ~[na:na]
		at java.base/java.util.concurrent.CompletableFuture.handle(CompletableFuture.java:2372) ~[na:na]
		at java.base/java.util.concurrent.CompletableFuture.handle(CompletableFuture.java:144) ~[na:na]
		at reactor.core.publisher.MonoCompletionStage$MonoCompletionStageSubscription.request(MonoCompletionStage.java:145) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.MonoFlatMapMany$FlatMapManyMain.onSubscribe(MonoFlatMapMany.java:141) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.MonoCompletionStage.subscribe(MonoCompletionStage.java:56) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.InternalFluxOperator.subscribe(InternalFluxOperator.java:68) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxDefer.subscribe(FluxDefer.java:54) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.Flux.subscribe(Flux.java:8891) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onComplete(FluxConcatArray.java:238) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxConcatArray.subscribe(FluxConcatArray.java:79) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.Flux.subscribe(Flux.java:8891) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onComplete(FluxConcatArray.java:238) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxFilterFuseable$FilterFuseableSubscriber.onComplete(FluxFilterFuseable.java:171) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxMapFuseable$MapFuseableConditionalSubscriber.onComplete(FluxMapFuseable.java:350) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onComplete(FluxConcatArray.java:209) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onComplete(Operators.java:2230) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onComplete(FluxMapFuseable.java:152) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onComplete(FluxSwitchIfEmpty.java:85) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxFilterFuseable$FilterFuseableSubscriber.onComplete(FluxFilterFuseable.java:171) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxHandleFuseable$HandleFuseableConditionalSubscriber.onComplete(FluxHandleFuseable.java:605) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxMapFuseable$MapFuseableConditionalSubscriber.onComplete(FluxMapFuseable.java:350) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxPeekFuseable$PeekFuseableConditionalSubscriber.onComplete(FluxPeekFuseable.java:595) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxPeekFuseable$PeekFuseableConditionalSubscriber.onComplete(FluxPeekFuseable.java:595) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxPeekFuseable$PeekFuseableConditionalSubscriber.onComplete(FluxPeekFuseable.java:595) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxPeekFuseable$PeekFuseableConditionalSubscriber.onComplete(FluxPeekFuseable.java:595) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxHandleFuseable$HandleFuseableConditionalSubscriber.onComplete(FluxHandleFuseable.java:605) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onComplete(FluxContextWrite.java:126) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxDoFinally$DoFinallySubscriber.onComplete(FluxDoFinally.java:128) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxPeekFuseable$PeekFuseableConditionalSubscriber.onComplete(FluxPeekFuseable.java:595) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxMapFuseable$MapFuseableConditionalSubscriber.onComplete(FluxMapFuseable.java:350) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxPeekFuseable$PeekFuseableConditionalSubscriber.onComplete(FluxPeekFuseable.java:595) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxPeekFuseable$PeekFuseableConditionalSubscriber.onComplete(FluxPeekFuseable.java:595) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxPeekFuseable$PeekFuseableConditionalSubscriber.onComplete(FluxPeekFuseable.java:595) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxPeekFuseable$PeekFuseableConditionalSubscriber.onComplete(FluxPeekFuseable.java:595) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxHandleFuseable$HandleFuseableConditionalSubscriber.onComplete(FluxHandleFuseable.java:605) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onComplete(FluxContextWrite.java:126) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxDoFinally$DoFinallySubscriber.onComplete(FluxDoFinally.java:128) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxPeekFuseable$PeekFuseableConditionalSubscriber.onComplete(FluxPeekFuseable.java:595) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxPublishOn$PublishOnConditionalSubscriber.doComplete(FluxPublishOn.java:1106) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxPublishOn$PublishOnConditionalSubscriber.checkTerminated(FluxPublishOn.java:1140) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxPublishOn$PublishOnConditionalSubscriber.runAsync(FluxPublishOn.java:996) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.publisher.FluxPublishOn$PublishOnConditionalSubscriber.run(FluxPublishOn.java:1079) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:84) ~[reactor-core-3.7.8.jar:3.7.8]
		at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:37) ~[reactor-core-3.7.8.jar:3.7.8]
		at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:317) ~[na:na]
		at java.base/java.util.concurrent.FutureTask.run(FutureTask.java) ~[na:na]
		at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[na:na]
		at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[na:na]
		at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[na:na]
		at java.base/java.lang.Thread.run(Thread.java:1583) ~[na:na]

### Expected Behavior

_No response_

### Steps To Reproduce

_No response_

### Environment

```markdown
Spring AI Alibaba version(s):
```

### Debug logs

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by 014-code**
## Summary

- Recover registered skill names emitted as unresolved tool calls by loading them through `read_skill`.
- Preserve real tools with the same name and existing unresolved responses.
- Carry dynamic callbacks and the agent runtime resolver into model request context so skill fallback cannot shadow tools resolved lazily.
- Activate grouped tools for recovered skills on the next model call.

## Testing

- `mvn -pl :spring-ai-alibaba-agent-framework -am -Dspotless.skip=true -Dtest=SkillsInterceptorEnhancementsTest,SkillToolCallInterceptorTest -Dsurefire.failIfNoSpecifiedTests=false -B test`
- `mvn -pl :spring-ai-alibaba-agent-framework -am -Dspotless.apply.skip=true -Dspotless.check.skip=false -DskipTests=true -Dmaven.source.skip=true -B spotless:check`

Fixes #4825

### CODE REVIEW

File: spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/interceptor/skills/SkillsInterceptor.java [hunk 1]
```diff
@@ -175,16 +177,40 @@ private List<SkillMetadata> extractReadSkills(List<Message> messages) {
...
+		if (ReadSkillTool.READ_SKILL.equals(toolCall.name())) {
+			return resolveSkillFromArguments(toolCall.arguments());
+		}
+		if (isRegisteredTool(request, toolCall.name())) {
+			return Optional.empty();
+		}
+		return skillRegistry.get(toolCall.name());
+	}
+
+	private boolean isRegisteredTool(ModelRequest request, String toolName) {
```
Comment (oss-sentinel-ai) [inline spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/interceptor/skills/SkillsInterceptor.java:199]: isRegisteredTool checks options.toolCallbacks and dynamicToolCallbacks but does not cover tools resolved at runtime by AgentToolNode.toolCallbackResolver or tools placed directly in RunnableConfig context under DYNAMIC_TOOL_CALLBACKS_METADATA_KEY. A real tool visible only through those sources could be misclassified as a skill on the next model call.

File: spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/interceptor/skills/SkillsInterceptor.java [hunk 2]
```diff
@@ -175,16 +177,40 @@ private List<SkillMetadata> extractReadSkills(List<Message> messages) {
...
 	}
 
+	private Optional<SkillMetadata> resolveSkillFromToolCall(AssistantMessage.ToolCall toolCall,
+			ModelRequest request) {
+		if (ReadSkillTool.READ_SKILL.equals(toolCall.name())) {
+			return resolveSkillFromArguments(toolCall.arguments());
+		}
+		if (isRegisteredTool(request, toolCall.name())) {
+			return Optional.empty();
+		}
```
Comment (oss-sentinel-ai) [inline spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/interceptor/skills/SkillsInterceptor.java:195]: resolveSkillFromToolCall falls back to skillRegistry.get(toolCall.name()) for any non-read_skill tool call that is not a registered tool. This message-scan path cannot see the unresolvedToolName metadata used by SkillToolCallInterceptor, so its safety depends entirely on isRegisteredTool being complete. Document or tighten the guard if possible.

File: spring-ai-alibaba-agent-framework/src/test/java/com/alibaba/cloud/ai/graph/agent/interceptors/SkillToolCallInterceptorTest.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+		ToolCallRequest request = request("allowed-tools-test");
+		ToolCallResponse response = interceptor.interceptToolCall(request,
+				ignored -> ToolCallResponse.success(request.getToolCallId(), request.getToolName(), "real tool result"));
+
+		assertEquals("real tool result", response.getResult());
+		assertEquals(ToolCallResponse.SUCCESS_STATUS, response.getStatus());
+	}
+
+	@Test
+	void preservesUnresolvedResponseForUnknownNonSkillTool() {
```
Comment (oss-sentinel-ai) [inline spring-ai-alibaba-agent-framework/src/test/java/com/alibaba/cloud/ai/graph/agent/interceptors/SkillToolCallInterceptorTest.java:101]: Add a test where the handler returns a real tool error (status error, but no unresolvedToolName metadata) to verify the interceptor does not swallow genuine tool failures.

File: spring-ai-alibaba-agent-framework/src/test/java/com/alibaba/cloud/ai/graph/agent/interceptors/SkillsInterceptorEnhancementsTest.java [hunk 4]
```diff
@@ -187,4 +187,32 @@ void groupedToolsSupplierIsResolvedOnEveryModelCall() {
 				.anyMatch(tool -> "extra_tool".equals(tool.getToolDefinition().name())));
 	}
 
+	@Test
+	void skillNameToolCallActivatesGroupedToolsAfterFallback() {
```
Comment (oss-sentinel-ai) [inline spring-ai-alibaba-agent-framework/src/test/java/com/alibaba/cloud/ai/graph/agent/interceptors/SkillsInterceptorEnhancementsTest.java:214]: Add a test where a real tool is registered under the same name as a skill (in request options/dynamicToolCallbacks) to confirm resolveSkillFromToolCall returns empty and does not activate grouped tools.

**Review by oss-sentinel-ai — COMMENTED** [review]
﻿### Summary
Adds a `SkillToolCallInterceptor` (registered via `SkillsAgentHook.getToolInterceptors()`) that recovers registered skill names the model emitted as unresolved tool calls by re-running them through `read_skill`, and updates `SkillsInterceptor` so such recovered skill reads activate the skill's grouped tools on the next model call. The approach is sound and non-blocking: the fallback keys off the framework's own `unresolvedToolName` metadata (produced only by `AgentToolNode`), requires `isError()` plus a matching tool name, so genuine tool failures are not swallowed; a real tool registered under the same name wins because the interceptor delegates to the base handler first; the hook contract wiring is correct and the interceptor name (`SkillToolCallFallback`) is unique, so no double-application risk. The main residual risk is the completeness of `SkillsInterceptor.isRegisteredTool` on the message-scan path (see inline comments). Tests cover the three core paths (recovery, real-tool-wins, unknown-tool-untouched) with a couple of edge cases worth adding.

### Findings
- **[Warning]** `SkillsInterceptor.java:198` — `isRegisteredTool` does not cover tools resolved at runtime by `AgentToolNode.toolCallbackResolver` or placed directly in `RunnableConfig` context under `DYNAMIC_TOOL_CALLBACKS_METADATA_KEY`; such a tool could be misclassified as a skill on the next model call.
- **[Warning]** `SkillsInterceptor.java:194` — the message-scan path (`resolveSkillFromToolCall`) cannot see the `unresolvedToolName` metadata, so its safety depends entirely on `isRegisteredTool` being complete; consider documenting or tightening this guard.
- **[Info]** `SkillToolCallInterceptorTest.java:84` — missing test: a real tool returning an error response without `unresolvedToolName` metadata must not be swallowed.
- **[Info]** `SkillsInterceptorEnhancementsTest.java:191` — missing test: a real tool registered under the same name as a skill must not activate grouped tools.

### Suggestions
1. If feasible, extend `isRegisteredTool` (or document explicitly) to cover runtime-resolver tools so a same-named skill cannot shadow a real tool between turns.
2. Consider adding the two missing edge-case tests above; they are cheap and directly lock in the "never mask real tool failures" guarantee.

---
*Automated review by github-manager-bot*

**Review by oss-sentinel-ai — APPROVED** [review]
### Summary
Re-review of follow-up commit `7f647ffa` ("guard skill fallback against runtime tools"). All four findings from the previous review are resolved:

- `isRegisteredTool` now covers runtime-resolved tools across all paths: dynamic callbacks on the request/options, context `DYNAMIC_TOOL_CALLBACKS_METADATA_KEY`, the interceptor-own `toolCallbackResolver`, and `TOOL_CALLBACK_RESOLVER_METADATA_KEY` propagated through `AgentLlmNode`/`RunnableConfig`.
- The message-scan path (`resolveSkillFromToolCall`) is gated by that complete check, so a real runtime-resolved tool can no longer be misclassified as a skill.
- Both requested edge-case tests are added (real tool failure without `unresolvedToolName` metadata is passed through; same-named real tools via dynamic callbacks / runtime context / resolver do not activate grouped tools), plus an end-to-end wiring test.
- Verified locally: agent-framework + graph-core compile cleanly and all 12 interceptor tests pass.

No new issues introduced by the follow-up commit. LGTM.

---
*Automated review by github-manager-bot*

**Review by oss-sentinel-ai — APPROVED** [review]
LGTM. The incremental commit since my last review adds a single focused test (`reactAgentBuilderPassesRuntimeResolverToLlmNode`) that pins the runtime `ToolCallbackResolver` wiring introduced by this PR.

Verified against the PR head:
- `Builder.resolver(ToolCallbackResolver)` exists in the shared agent `Builder` and stores the resolver.
- `DefaultBuilder` forwards it to the LLM node via `.toolCallbackResolver(this.resolver)`.
- The reflection targets used by the test match the current code: `ReactAgent.llmNode` (declared on `ReactAgent`) and `AgentLlmNode.toolCallbackResolver`.
- All required imports are already present in the test class; no production code changed in this commit.

Nice regression coverage for direct skill tool call resolution.

---
*Automated review by github-manager-bot*
